larpandoracontent/LArObjects/LArAdaBoostDecisionTree.cc

```diff
@@ -0,0 +1,299 @@
+
+AdaBoostDecisionTree::Node::Node(const int &nodeID, const int &parentNodeID, const int &leftChildNodeID, const int &rightChildNodeID, const bool &isLeaf, const double &threshold, const int &variableID, const bool outcome) : 
+    m_nodeID(nodeID),
+    m_parentNodeID(parentNodeID),
+    m_leftChildNodeID(leftChildNodeID),
+    m_rightChildNodeID(rightChildNodeID),
+    m_isLeaf(isLeaf),
+    m_threshold(threshold),
+    m_variableID(variableID),
+    m_outcome(outcome)
+{
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+AdaBoostDecisionTree::WeakClassifier::WeakClassifier(IDToNodeMap &idToNodeMap, const double &weight, const int &treeID) : 
+    m_idToNodeMap(idToNodeMap),
+    m_weight(weight),
+    m_treeID(treeID)
+{
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+bool AdaBoostDecisionTree::WeakClassifier::Predict(const DoubleVector &features) const 
+{   
+    return this->EvaluateNode(0, features); 
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+bool AdaBoostDecisionTree::WeakClassifier::EvaluateNode(const int &nodeID, const DoubleVector &features) const 
+{
+    const Node *pActiveNode(nullptr);
+
+    if (m_idToNodeMap.find(nodeID) != m_idToNodeMap.end())
+    {
+        pActiveNode = m_idToNodeMap.at(nodeID);
+    }
+
+    if (pActiveNode->IsLeaf())
+        return pActiveNode->GetOutcome();
+
+    if (features.at(pActiveNode->GetVariableID()) <= pActiveNode->GetThreshold())
+    {
+        return this->EvaluateNode(pActiveNode->GetLeftChildNodeID(), features);
+    }
+    else 
+    {
+        return this->EvaluateNode(pActiveNode->GetRightChildNodeID(), features);
+    }
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+AdaBoostDecisionTree::StrongClassifier::StrongClassifier(WeakClassifiers &weakClassifiers) : 
+    m_weakClassifiers(weakClassifiers)    
+{
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+double AdaBoostDecisionTree::StrongClassifier::Predict(const DoubleVector &features) const
+{
+    double score(0.0);
+    double weights(0.0);
+
+    for (const WeakClassifier *pWeakClassifier : m_weakClassifiers)
+    {
+        weights += pWeakClassifier->GetWeight();
+
+        if (pWeakClassifier->Predict(features))
+	{
+             score += pWeakClassifier->GetWeight();
+        }
+        else
+        {
+            score -= pWeakClassifier->GetWeight();
+        }
+    }
+
+    score /= weights;
```

Comment:
Care, could divide by zero here if e.g. `m_weakClassifiers` is empty or classifier weights are zero.

Reply:
Check added and exception thrown if weights sum zero.  This would be equivalent to m_weakClassifiers being empty so no additional check on this implemented.